### PR TITLE
Send healthcheck after lds request has been sent

### PR DESCRIPTION
**Pull request by hzxuzhonghu**
**Please provide a description of this PR:**

Fix https://github.com/istio/istio/issues/35860

### CODE REVIEW

**Comment by ramaraochavali** [comment]
LGTM

File: pilot/pkg/xds/ads.go [hunk 1]
```diff
@@ -167,6 +167,10 @@ func (s *DiscoveryServer) receive(con *Connection) {
 		}
 		// This should be only set for the first request. The node id may not be set - for example malicious clients.
 		if firstRequest {
+			// probe happens before envoy sends first xDS request
+			if req.TypeUrl == v3.HealthInfoType {
+				continue
```
[the remark this answers is not in the document]
Comment (hzxuzhonghu) [reply pilot/pkg/xds/ads.go:172]: Warn sgtm


>We also might want to allow it if the Node info is there for some reason
We are quite sure it is not allowed healthcheck to be the first request.

File: pkg/istio-agent/xds_proxy.go [hunk 2]
```diff
@@ -337,6 +337,10 @@ func (p *XdsProxy) handleStream(downstream adsStream) error {
 				}
 				return
 			}
+			if req.TypeUrl == v3.HealthInfoType && !initialRequestsSent {
```
[the remark this answers is not in the document]
Comment (hzxuzhonghu) [reply pkg/istio-agent/xds_proxy.go:340]: Right, it is weird to have two separate threads sending xds requests. I can merge them

**Comment by hzxuzhonghu** [comment]
@howardjohn another look

File: pilot/pkg/xds/ads.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (howardjohn) [reply pilot/pkg/xds/ads.go:173]: Seems valid to me but I think its a very very niche case and the client code doesn't support it anyways.

File: pkg/istio-agent/xds_proxy.go [hunk 3]
```diff
@@ -449,10 +409,55 @@ func (p *XdsProxy) HandleUpstream(ctx context.Context, con *ProxyConnection, xds
 }
 
 func (p *XdsProxy) handleUpstreamRequest(con *ProxyConnection) {
+	// Handle downstream xds
```
Comment (howardjohn) [inline pkg/istio-agent/xds_proxy.go:412]: Having a function named `handleUpstreamRequest` that does `Handle downstream xds` seems misleading?

File: pkg/istio-agent/xds_proxy.go [hunk 4]
```diff
@@ -449,10 +409,55 @@ func (p *XdsProxy) HandleUpstream(ctx context.Context, con *ProxyConnection, xds
...
+			}
+		}
+	}()
+
 	defer con.upstream.CloseSend() // nolint
 	for {
 		select {
 		case req := <-con.requestsChan:
+			if req.TypeUrl == v3.HealthInfoType && !initialRequestsSent.Load() {
+				// only send healthcheck probe after LDS request has been sent
```
Comment (howardjohn) [inline pkg/istio-agent/xds_proxy.go:458]: can we add a test to make sure it is actually correctly sent after LDS? Worried we may drop the update

File: pkg/istio-agent/xds_proxy.go [hunk 5]
```diff
@@ -449,10 +409,55 @@ func (p *XdsProxy) HandleUpstream(ctx context.Context, con *ProxyConnection, xds
...
+					})
+				}
+				// Fire of a configured initial request, if there is one
+				p.connectedMutex.RLock()
+				initialRequest := p.initialRequest
+				if initialRequest != nil {
+					con.sendRequest(initialRequest)
+				}
+				p.connectedMutex.RUnlock()
+				initialRequestsSent.Store(true)
```
Comment (howardjohn) [inline pkg/istio-agent/xds_proxy.go:448]: I think there is a race condition that will lead to droppped requests:

```
a) con.SendRequest
a) load p.initialRequest, it is nil. Unlock connectedMutex
b) PersistRequest, store initialRequest and send on req chan
c) read from requestChan. initialRequestsSent=false, do nothing
a) `initialRequestsSent.Store(true)`
```

Letter == the goroutine.

I think we can just move this Store into the connectedMutex lock?

File: pkg/istio-agent/xds_proxy.go (hunk 3 above)
Comment (hzxuzhonghu) [reply pkg/istio-agent/xds_proxy.go:412]: maybe should recomment it

File: pkg/istio-agent/xds_proxy.go (hunk 5 above)
Comment (hzxuzhonghu) [reply pkg/istio-agent/xds_proxy.go:448]: good catch, i think move the initialRequestsSent.Store(true) before the `con.sendRequest(initialRequest)` should be ok.

File: pkg/istio-agent/xds_proxy.go (hunk 4 above)
Comment (hzxuzhonghu) [reply pkg/istio-agent/xds_proxy.go:458]: ok, will add to it

Comment (hzxuzhonghu) [reply pkg/istio-agent/xds_proxy.go:458]: done

**Review by howardjohn — APPROVED** [review]
ty!

**Pull request merged**